Replace theme add-on with toolbar and global args

tylersticka opened the pull request:
## Overview

After upgrading Storybook (#1214), our themes add-on became available in Docs view but was not functional. I attempted to use he add-on's `withThemes` decorator, but it does not appear to be compatible with inlined HTML stories. So this PR replaces it with a custom toolbar using the Toolbar add-on that ships with Storybook's essential add-ons.

This also helps us overcome an issue where we couldn't reliably configure a theme for a specific story. Now, we can specify a `theme` parameter corresponding to the desired theme class and the story will gain that theme. That means we can remove the `fill-canvas` template we were using as kind of a hack.

## Screenshots

<img width="1050" alt="Screen Shot 2021-05-14 at 2 42 32 PM" src="https://user-images.githubusercontent.com/69633/118334581-bfd51500-b4c2-11eb-9023-81725e189e7c.png">

<img width="850" alt="Screen Shot 2021-05-14 at 2 42 54 PM" src="https://user-images.githubusercontent.com/69633/118334586-c06dab80-b4c2-11eb-8779-ae8af19efd0c.png">

<img width="551" alt="Screen Shot 2021-05-14 at 2 43 07 PM" src="https://user-images.githubusercontent.com/69633/118334588-c1064200-b4c2-11eb-8243-ee5084d76eff.png">

## Testing

On the deploy preview…

- Try the theme toolbar in a Canvas view
- Try the theme toolbar in a Docs view

Paul-Hebert commented:
This is looking really good Tyler! I noticed a few oddities:

### Broken Docs Pages

Some of our stories have broken docs pages with this change. For example, when I view one of our Prism docs pages I get a console error. Switching to the Canvas view shows that error in the canvas

<img width="1440" alt="Screen Shot 2021-05-20 at 4 03 38 PM" src="https://user-images.githubusercontent.com/5798536/119059628-8ea68a00-b985-11eb-8d6a-86b05eabe57a.png">

### Getting Stuck in a Theme

I'm not exactly sure why this happens but it seems like on the Input docs page specifically I can get stuck in a theme:

![theme-stuck](https://user-images.githubusercontent.com/5798536/119059658-a0882d00-b985-11eb-97db-93ddcda98fe7.gif)

### Bunch

For some reason the Bunch example disappears on the canvas in dark mode. Upping the z-index on `.o-bunch` fixes it. I'm not sure if this is a regression or an existing issue.

<img width="1201" alt="Screen Shot 2021-05-20 at 4 00 14 PM" src="https://user-images.githubusercontent.com/5798536/119059445-2c4d8980-b985-11eb-85f0-0cdb66dcd350.png">

### Table Theme Issues

This is unrelated to this PR, but the table has some issues in dark mode (low contrast caption color, incorrect striped color.)

What's the best place for me to report that? A new issue?

---

Other than that it's looking great! Let me know if you need any more info on any of this!

tylersticka commented:
These are all great catches, @Paul-Hebert , thank you for testing so thoroughly!

spaceninja commented on src/components/button/button.stories.mdx:
```diff
@@ -35,10 +34,7 @@ Some buttons may not require as much emphasis as others. The `c-button--secondar
-    parameters={{ themes: { disable: true }, paddings: { disabled: true } }}
-  >
-    {(args) => stylesDarkDemo(args)}
+  <Story name="Styles (Dark)" parameters={{ theme: 't-dark' }}>
```
Should this be `dark` or `t-dark`? You used `dark` in `themes.stories.mdx`.

tylersticka replied:
It should be `t-dark`, the `dark` was me passing a noun for the demonstrated theme to the theme demo specifically.

spaceninja commented on src/design/themes.stories.mdx:
```diff
@@ -23,15 +13,17 @@ Themes may be applied to a containing element to change that content's appearanc
 
 <Canvas>
-  <Story name="Dark" args={{ theme: 'dark' }}>
+  <Story name="Dark" parameters={{ theme: 't-dark' }} args={{ theme: 'dark' }}>
```
Should this be `dark` or `t-dark`? You used `t-dark` in `button.stories.mdx`.

tylersticka replied:
It's kind of confusing, but I'm using `args` to tell the demo which theme is applied, and `parameters` to tell the decorator which theme to use on the container.

Storybook's weird.

spaceninja commented on src/design/themes.stories.mdx:
```diff
@@ -23,15 +13,17 @@ Themes may be applied to a containing element to change that content's appearanc
 
 <Canvas>
-  <Story name="Light">{demoStory.bind({})}</Story>
+  <Story name="Light" args={{ theme: 'light' }}>
```
Should this be `light` or `t-light`?

tylersticka replied:
In this case this is me telling the demo which theme is applied so it can modify its text and stuff accordingly, so even though it looks really close to the parameters it's not actually the same thing. I'm hoping this will only happen in the _theme_ demos.

tylersticka commented:
@Paul-Hebert I just pushed some updates that _may_ address the issues you've observed...

> Broken Docs Pages

I couldn't reproduce this error, but I added extra checks to make sure elements exist before using `querySelector` or `classList` in hopes that it will prevent this from occurring.

> Getting Stuck in a Theme

I also couldn't reproduce this error, but I noticed a different console error relating to my use of `value: undefined` which could, theoretically, get the toolbar stuck. I changed to `null`, which seemed to eliminate the warning... but it may have nothing to do with what you were seeing.

> Bunch

I _could_ reproduce this one! I think I fixed it by applying the theme class to `document.body` instead of `document.documentElement`.

> Table Theme Issues

I've created an issue to resolve this:  #1229

spaceninja commented:
I'm seeing the same problem @Paul-Hebert saw on the Input page — changing the theme to dark on the Input docs page doesn't change anything.

tylersticka commented:
> I'm seeing the same problem @Paul-Hebert saw on the Input page — changing the theme to dark on the Input docs page doesn't change anything.

Okay, _now_ I can reproduce. Not sure what I'm doing differently now. I think this may have something to do with inline versus iframed stories.

Thanks, @spaceninja! Sorry for the runaround.

spaceninja commented:
<img width="998" alt="Screen Shot 2021-05-20 at 5 04 29 PM" src="https://user-images.githubusercontent.com/257309/119063232-733f7d00-b98d-11eb-89a3-01ec597c3c6b.png">

I'm not sure if this is a bug, because I'm not sure it's a real-world scenario, but if I apply the global dark theme on the Themes docs page, the nested dark card inside the light theme (that is now dark theme) remains dark. I would expect that if one changes, the other does also, but feel free to disregard if this isn't a situation a real user could end up in.

tylersticka commented:
> I'm not sure if this is a bug, because I'm not sure it's a real-world scenario, but if I apply the global dark theme on the Themes docs page, the nested dark card inside the light theme (that is now dark theme) remains dark. I would expect that if one changes, the other does also, but feel free to disregard if this isn't a situation a real user could end up in.

Good catch. I think at the very least we could force them to stay the same theme regardless, since this is kind of a special case and someone could (as you suggest) swap themes and then head over here.

spaceninja commented:
> I think at the very least we could force them to stay the same theme regardless

Could we disable the theme control on that page?

tylersticka commented:
Gonna have to pick this up tomorrow, but there may be a major blocker to getting this working with iframed docs: https://github.com/storybookjs/storybook/issues/13444

It doesn't even seem like the iframed docs are responding to a change in that toolbar, but I'd like to take a new look with fresh eyes tomorrow.

tylersticka commented:
> Could we disable the theme control on that page?

Probably!

tylersticka commented:
I think this bug may be more closely related to the issue we're seeing on Input: https://github.com/storybookjs/storybook/issues/14477

gerardo-rodriguez commented:
@tylersticka I can't tell the state of this PR at a glance. Are you in the process of addressing feedback or is the PR in a ready state for another review. Let me know how I can help! 🙂

tylersticka commented:
> I can't tell the state of this PR at a glance. Are you in the process of addressing feedback or is the PR in a ready state for another review. Let me know how I can help! 🙂

@gerardo-rodriguez I'm in the process of addressing feedback at the moment! I just removed the open request for Dev review, sorry about that!

tylersticka commented:
Okay, I think this is ready for re-review, with a few caveats:

- I can't easily disable the toolbar for specific story files. I've overtly set the theme on both theme stories to address [this comment](https://github.com/cloudfour/cloudfour.com-patterns/pull/1219#issuecomment-845560550).
- I couldn't find a solution so that themes would work for non-inlined stories in Docs mode. I've updated the custom decorator comments to include more information about why this is a struggle.

Both issues are solvable if we wanted to expand this from a custom toolbar and decorator to a full add-on, which would give us the ability to assign more complicated actions to the toolbar items and would allow us to force-refresh stories in contexts like the docs mode. I haven't done this because...

- The remaining issues are known and documented in Storybook with some dev effort already allocated to them. It's likely that these issues will be trivial to address in a future Storybook update.
- Every single existing add-on for this that's compatible with Storybook 6.x has _more_ bugs than this implementation (including what we were using before, which didn't work in Docs mode at all for HTML or Web Component "frameworks"), which suggests to me that this is a non-trivial problem to fully solve.
- The majority of our pattern stories are inlined, as that is the default. Of the seven patterns that are not inlined, three are layout objects, two are theme containers themselves (Sky Nav and Cloud Cover), one is only optimized for a light theme (Ground Nav), and the last pattern (Input) can be safely inlined once we have more form layouts to better visualize the alignment of inputs, labels and buttons (see #1088 and #1216).

TL;DR: This isn't perfect, but it's a lot better than our current solution, and is likely to improve more as our dependencies update and our project evolves. But let me know if you disagree!

gerardo-rodriguez commented on .storybook/theme-decorator.js:
```diff
@@ -0,0 +1,65 @@
+      const storyElement = document.querySelector(`#story--${context.id}`);
+      // Only proceed if we actually found the relevant story.
+      if (storyElement) {
+        const previewElement = storyElement.closest('.docs-story');
+        updateTheme(previewElement, theme);
```
I don't yet have an understanding of story elements and their contents. Is it a guaranty that there will always exist a `.docs-story` or do we need to check that `previewElement` is not `null` before attempting to update the theme? 🤔 ¯\\\_(ツ)_/¯

Paul-Hebert replied:
This seems like a good safety check given that the underlying Storybook markup is outside our control

tylersticka replied:
The `updateTheme` function does check that the element is truthy and has a `classList` property, so I don't think another check before calling it is necessary.

Paul-Hebert reviewed (approved):
> TL;DR: This isn't perfect, but it's a lot better than our current solution, and is likely to improve more as our dependencies update and our project evolves. But let me know if you disagree!

That makes sense to me :+1: I'm cool with merging this in, though agree with Gerardo that it would be nice to add a safety check to confirm the `.story-docs` element exists before interacting with it.
